clang-tools-extra/clangd/refactor/tweaks/InlineConceptRequirement.cpp

```diff
@@ -0,0 +1,276 @@
+
+  template <typename T, typename NodeKind>
+  static auto findNode(const SelectionTree::Node &Root)
+      -> std::tuple<const T *, const SelectionTree::Node *>;
+
+  template <typename T>
+  static auto findExpression(const SelectionTree::Node &Root)
+      -> std::tuple<const T *, const SelectionTree::Node *> {
+    return findNode<T, Expr>(Root);
+  }
+
+  template <typename T>
+  static auto findDeclaration(const SelectionTree::Node &Root)
+      -> std::tuple<const T *, const SelectionTree::Node *> {
+    return findNode<T, Decl>(Root);
+  }
+};
+
+REGISTER_TWEAK(InlineConceptRequirement)
+
+bool InlineConceptRequirement::prepare(const Selection &Inputs) {
+  if (!Inputs.AST->getLangOpts().CPlusPlus20) {
+    return false;
+  }
+
+  const auto *Root = Inputs.ASTSelection.commonAncestor();
+  if (!Root) {
+    return false;
+  }
+
+  const SelectionTree::Node *ConceptSpecializationExpressionTreeNode;
+  std::tie(ConceptSpecializationExpression,
+           ConceptSpecializationExpressionTreeNode) =
+      findExpression<ConceptSpecializationExpr>(*Root);
+  if (!ConceptSpecializationExpression) {
+    return false;
+  }
+
+  // Only allow concepts that are direct children of function template
+  // declarations or function declarations. This excludes conjunctions of
+  // concepts which are not handled.
+  const auto *ParentDeclaration =
+      ConceptSpecializationExpressionTreeNode->Parent->ASTNode.get<Decl>();
+  if (!isa_and_nonnull<FunctionTemplateDecl>(ParentDeclaration) &&
+      !isa_and_nonnull<FunctionDecl>(ParentDeclaration)) {
+    return false;
+  }
+
+  const FunctionTemplateDecl *FunctionTemplateDeclaration =
+      std::get<0>(findDeclaration<FunctionTemplateDecl>(*Root));
+  if (!FunctionTemplateDeclaration) {
+    return false;
+  }
+
+  auto TemplateArguments =
+      ConceptSpecializationExpression->getTemplateArguments();
+  if (TemplateArguments.size() != 1) {
+    return false;
+  }
+
+  auto TemplateParameterIndex =
+      getTemplateParameterIndexOfTemplateArgument(TemplateArguments[0]);
+  if (!TemplateParameterIndex) {
+    return false;
+  }
+
+  TemplateTypeParameterDeclaration = dyn_cast_or_null<TemplateTypeParmDecl>(
+      FunctionTemplateDeclaration->getTemplateParameters()->getParam(
+          *TemplateParameterIndex));
+  if (!TemplateTypeParameterDeclaration->wasDeclaredWithTypename()) {
+    return false;
+  }
+
+  RequiresToken =
+      findToken(Inputs.AST, FunctionTemplateDeclaration->getSourceRange(),
+                tok::kw_requires);
+  if (!RequiresToken) {
+    return false;
+  }
+
+  return true;
+}
+
+Expected<Tweak::Effect>
+InlineConceptRequirement::apply(const Selection &Inputs) {
+  auto &Context = Inputs.AST->getASTContext();
+  auto &TokenBuffer = Inputs.AST->getTokens();
+
+  tooling::Replacements Replacements{};
+
+  if (auto Err =
+          Replacements.add(generateTemplateParameterReplacement(Context))) {
+    return std::move(Err);
```

Comment:
```suggestion
    return std::move(Err);
```
`Err` is treated as a an rvalue in this case anyway. This potentially pessimizes NRVO potenitally.

Reply:
Same below, but there it is RVO.

Reply:
going through the other tweaks they return errors with `return std::move(Err);` or returning an error with a comment `return error("An error occured");`.

Reply:
I'd keep it without the move unless they ask us to change it. The compiler can infer this on it's own. I suspect they wrote this before C++17.

https://en.cppreference.com/w/cpp/language/copy_elision

Reply:
Removed moves where possible